### PR TITLE
Add function get_device

### LINKED ISSUE #268: Adding a function to get device from array (type)?
It would be nice to have a function that returns a KernelAbstractions.jl device given an array (type), e.g.

```
get_device(A::AbstractArray) = get_device(typeof(A))

get_device(::Type{<:AbstractArray}) = CPUDevice()
get_device(T::Type{<:AbstractGPUArray}) = throw(ArgumentError("NoKernelAbstractions.Device type defined for arrays of type $(T.name.name)"))

get_device(::Type{<:CuArray}) = CUDADevice()
get_device(::Type{<:AbstractCuSparseArray}) = CUDADevice()

get_device(::Type{<:ROCArray}) = ROCDevice()
```

so that users can write device-agnostic code like

```
function foo(A::AbstractArray)
    A = ones(1024, 1024)
    device = KernelAbstractions.get_device(A)
    kernel = mul2(device, 16)
    event = kernel(A, ndrange=size(A))
    ...
```

Would that fit into the KernelAbstractions concept?

### PR DESCRIPTION
**Pull request by oschulz**
Closes #268

### CODE REVIEW

**Comment by oschulz** [comment]
But as for the example - silly me, we can just do

```julia
    device = KernelAbstractions.get_device(A)
    n = device isa GPU ? 256 : 4
    kernel! = naive_transpose_kernel!(device, n)
```

Very literate, I think. :-)

**Comment by vchuravy** [comment]
bors try

**Comment by oschulz** [comment]
Is bors hanging?

**Comment by jpsamaroo** [comment]
ROCm CI got hung after I updated the local buildkite repo; I've restarted the runners, which are now processing jobs.

**Comment by oschulz** [comment]
Arg, forgot some tests deps ... fixed now.

**Comment by oschulz** [comment]
Go bors, go :-)

**Comment by oschulz** [comment]
Ok, CUDA and ROC fail, but it looks like for unrelated reasons, correct?

**Comment by jpsamaroo** [comment]
Sorry, forgot to bump the AMDGPU version after Valentin fixed this issue. Wait 30 mins, then bump the ROCKernels dep of AMDGPU.jl to 0.2.14.

**Comment by oschulz** [comment]
Bump - can someone kick bors?

**Comment by jpsamaroo** [comment]
bors try

**Comment by oschulz** [comment]
These bors error are unrelated to this PR, right?

**Comment by jpsamaroo** [comment]
This is an AMDGPU bug for sure. @vchuravy should I push some code to allow the build to fail on the `using` statements, or should we stall this PR until we have ROCm LLVM 12 support working in Yggdrasil?

**Comment by vchuravy** [comment]
Let's let this fail. I also think we need to decouple CI for ROCKernels and CUDAKernels

**Comment by vchuravy** [comment]
No we will still need CI back to working, before we can merge this.

**Comment by oschulz** [comment]
Sure, of course.

**Comment by oschulz** [comment]
Is CI back? I saw there was a new release.

**Comment by jpsamaroo** [comment]
Not yet, but once https://github.com/JuliaPackaging/Yggdrasil/pull/3776 is merged (and it should be ready to go now), I can post PRs for all the rest of the ROCm 4.2.0 packages, and then post and merged an AMDGPU PR to use them on 1.7. Maybe another week if all goes well?

**Comment by oschulz** [comment]
Thanks @jpsamaroo !

**Comment by vchuravy** [comment]
Can you rebase on master? Just made CI work again

**Comment by oschulz** [comment]
Thanks - rebase done, fingers crossed. :-)

**Comment by jpsamaroo** [comment]
@oschulz we don't currently support sparse arrays in AMDGPU; I would disable that test on ROCm CI for now.

**Comment by oschulz** [comment]
Ok disabled the test, can you run CI again?

**Comment by jpsamaroo** [comment]
I think I need to get https://github.com/JuliaGPU/AMDGPU.jl/pull/187 working and merged before ROCm CI is going to be ready.

**Comment by oschulz** [comment]
Sorry, made a silly mistake - can you run CI again?

**Comment by oschulz** [comment]
I think the remaining test failures are unrelated to this PR (correct?).

**Comment by oschulz** [comment]
I saw https://github.com/JuliaGPU/AMDGPU.jl/pull/187 got merged with a new AMDGPU.jl release. Should we try bors again?

**Comment by oschulz** [comment]
@jpsamaroo , could you kick bors again, with the new AMD fixes?

**Comment by jpsamaroo** [comment]
@oschulz you would need to rebase on https://github.com/JuliaGPU/KernelAbstractions.jl/pull/288, although that PR isn't yet CI-green.

**Comment by oschulz** [comment]
>  rebase on #288, although that PR isn't yet CI-green.

Ah Ok - maybe easiest to just wait until #288 is in, then?

File: Project.toml [hunk 1]
```diff
@@ -6,7 +6,10 @@ version = "0.8.0-dev"
 [deps]
 Adapt = "79e6a3ab-5dfb-504d-930d-738a2a938a0e"
 InteractiveUtils = "b77e0a4c-d291-57a0-90e8-8db25a27a240"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
 MacroTools = "1914dd2f-81c6-5fcd-8719-6d5c9610ff09"
+SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
+SpecialFunctions = "276daf66-3868-5448-9aa4-cd146d93841b"
```
Comment (vchuravy) [inline Project.toml:12]:
```suggestion
```

File: lib/ROCKernels/Project.toml [hunk 2]
```diff
@@ -15,5 +15,6 @@ StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
 AMDGPU = "0.3.2"
 Adapt = "3.0"
 KernelAbstractions = "0.8"
+SpecialFunctions = "0.10, 1.0, 2.0"
```
Comment (vchuravy) [inline lib/ROCKernels/Project.toml:18]:
```suggestion
```

**Comment by oschulz** [comment]
I saw you made some changes @vchuravy , do I need to do anything in addition?

(Am I correct that the current test failures are unrelated to this PR?)

**Comment by vchuravy** [comment]
The ROCM failure looks related, only Julia nightly is currently expected to fail.

**Comment by vchuravy** [comment]
Thank you Oliver!

**Comment by oschulz** [comment]
Thank you Valentin!

**Pull request merged**